### ISSUE TITLE
`Interchange.to_openmm_simulation()`

### ISSUE THREAD

**Issue by Yoshanuikabundi** [openforcefield/openff-interchange#713]
Don't worry about this until after the meeting, just wanted to put this down so I don't forget.

**Description**
In https://github.com/openforcefield/openff-docs/pull/5 we discussed an example that goes from a SMILES string to a prepared set of OpenMM system/topology/positions objects, and how it would be nice to be able to actually run the simulation as a part of this. I was wondering if it would be possible to one day include an `Interchange.to_openmm_simulation()` method, which would take an `openmm.Integrator` as an argument and produce an OpenMM `Simulation`, ready to go. So a bare-bones SMILES-to-simulation would go from:

```python
from openff.toolkit import Molecule, Topology, ForceField
from openff.interchange import Interchange
import openmm

paracetamol = Molecule.from_smiles("CC(=O)NC1=CC=C(C=C1)O")
paracetamol.generate_conformers(n_conformers=1)

topology = Topology.from_molecules([paracetamol])

force_field = ForceField("openff-2.0.0.offxml")

interchange = Interchange.from_smirnoff(force_field, topology)

openmm_system = interchange.to_openmm()
openmm_topology = interchange.to_openmm_topology()
openmm_positions = interchange.positions.to_openmm()

simulation = openmm.app.simulation.Simulation(
    openmm_topology, 
    openmm_system, 
    openmm.VerletIntegrator(2.0 * openmm.unit.femtosecond),
)
simulation.context.setPositions(openmm_positions)
simulation.minimizeEnergy()
simulation.context.getState(getEnergy=True).getPotentialEnergy()
```

to 

```python
from openff.toolkit import Molecule, Topology, ForceField
from openff.interchange import Interchange
import openmm

paracetamol = Molecule.from_smiles("CC(=O)NC1=CC=C(C=C1)O")
paracetamol.generate_conformers(n_conformers=1)

topology = Topology.from_molecules([paracetamol])

force_field = ForceField("openff-2.0.0.offxml")

interchange = Interchange.from_smirnoff(force_field, topology)

simulation = interchange.to_openmm_simulation(
    openmm.VerletIntegrator(2.0 * openmm.unit.femtosecond),
)
simulation.minimizeEnergy()
simulation.context.getState(getEnergy=True).getPotentialEnergy()
```

Missing positions would just leave the context unset. Users would provide the constructed integrator, so we wouldn't have to add any special handling for it. This would simplify almost all our examples, and for many users would just remove intermediate objects that they only need to construct a `Simulation`. The existing methods would remain for users who need or prefer them. We could have a `**kwargs` argument to pass any other arguments on to the `Simulation` constructor, which we could link to in the docstring. Implementation would be something like:

```python
def to_openmm_simulation(self, integrator, **kwargs):
    simulation = openmm.Simulation(
        interchange.to_openmm(),
        interchange.to_openmm_topology(),
        integrator,
        **kwargs,
    )

    positions = interchange.positions
    if positions is not None:
        simulation.context.setPositions(positions.to_openmm())

    return simulation
```

We might have to investigate if its safe to modify a system after constructing a simulation, but the [attribute](https://github.com/openmm/openmm/blob/ed049af1b34cab4dd47f20dcfddb99bcad6c7913/wrappers/python/openmm/app/simulation.py#L84-L88) is public and there's no warning in the [API documentation](http://docs.openmm.org/8.0.0/api-python/generated/openmm.app.simulation.Simulation.html) so hopefully that's OK. If not, users would have to use the existing machinery to add things like pressure coupling, so this would be limited in value.

**Comment by mattwthompson** [comment]
I'm fairly strongly in support of this; reducing boilerplate code is good and there should be limited callbacks between OpenMM and here. There shouldn't be [much in the way of kwargs](http://docs.openmm.org/latest/api-python/generated/openmm.app.simulation.Simulation.html#openmm.app.simulation.Simulation.__init__) to pass to `Simulation()`. The only detail I would have proposed differently is setting a default value of the integrator. That can be discussed elsewhere.

Do you think it'd be useful to have `Interchange.to_openmm_system` as an alias to `.to_openmm`? I'm not necessarily suggesting dropping the latter, but I'd be open to it. Exporting out to three or four different OpenMM objects makes the name kinda weird.

If you think it's not a terrible idea to add some vanilla energy minimization to the API, that would be straightforward enough as well. We did [something similar](https://github.com/mosdef-hub/mbuild/blob/0.16.1/mbuild/compound.py#L2006) in mBuild years back.


Neither here nor there, I got tired of this pattern after the eleventieth billionth zillion time typing it out:

```python
paracetamol = Molecule.from_smiles("CC(=O)NC1=CC=C(C=C1)O")
paracetamol.generate_conformers(n_conformers=1)
```
and I'd love to have an argument in `Molecule` class methods that does this by default 🙃 

```python
paracetamol = Molecule.from_smiles("CC(=O)NC1=CC=C(C=C1)O", give_me_conformers_pretty_please=True)
```

**Comment by Yoshanuikabundi** [comment]
> Do you think it'd be useful to have `Interchange.to_openmm_system` as an alias to `.to_openmm`?

Yes, I think this would be a clearer name and more consistent with how other exports are named. `to_openmm` always felt to me like it implied that it would produce everything you need for an OpenMM simulation. `to_openmm_positions` might be worth implementing too, even though `interchange.positions.to_openmm()` does the same thing.

> and I'd love to have an argument in Molecule class methods that does this by default

or even just make `generate_conformers` return `self` so that `Molecule.from_smiles(...).generate_conformers(n_conformers=1)` works... and make the first argument to generate_conformers `n_conformers` because that's almost always the one people want to change. So `mol = Molecule.from_smiles(...).generate_conformers(1)`

**Comment by Yoshanuikabundi** [comment]
> If you think it's not a terrible idea to add some vanilla energy minimization to the API, that would be straightforward enough as well. We did [something similar](https://github.com/mosdef-hub/mbuild/blob/0.16.1/mbuild/compound.py#L2006) in mBuild years back.

Now that I think about it, this might be valuable for users who want identical starting structures to compare dynamics in different engines or something. Not sure that any such users exist but its possible! It also saves the user from defining an integrator that they'll never use if all they want is to energy minimise. It could even just act on the positions in the Interchange.

**Comment by mattwthompson** [comment]
This was implemented in #725 and included in 0.3.5
